Write up on schema evolution (#1960)
hazelcast/hz-docs: docs/modules/serialization/pages/schema-evolution.adoc

== Introducing schema evolution
:description: This topic introduces schema evolution in Hazelcast Compact serialization and explains how to safely evolve object schemas across rolling upgrades, clients, and persisted clusters without data loss or downtime. It covers backward and forward compatibility, deterministic writes, versioned maps, MapStore and persistence considerations, SQL and index evolution, and best practices for zero-downtime migrations in distributed environments.

{description}

> Schema evolution refers to the ability of Hazelcast to handle changes in the structure of serialized objects without requiring downtime, full data reloads, or strict synchronization across cluster members and clients.

Reviewer: what does this mean? `strict synchronization across cluster members and clients`

In distributed systems, data structures evolve as applications change. You might need to add a field to a class, widen a type, or modify a nested record. In traditional serialization formats, these changes can cause deserialization failures or inconsistent reads when old and new members coexist. Hazelcast Compact serialization resolves these challenges through **schema evolution**, enabling smooth coexistence of multiple schema versions for the same object type across the cluster.

Schema evolution in Hazelcast focuses on three goals:

1. **Backward compatibility:** New members can read data written by old members.

> 2. **Forward compatibility:** Old members can read data written by new members, ignoring unknown fields.

Author: I meant to state a "definition" of forward compatibility. When FC can't be guaranteed is because of a breaking change

3. **Operational continuity:** Schemas evolve without invalidating stored data, map entries, or persisted state.

> Hazelcast achieves this through Compact serialization, which stores object schemas as independent metadata and identifies each version using a unique 64-bit fingerprint. When an object changes, Hazelcast detects the new schema automatically, assigns it a distinct fingerprint, and propagates it across the cluster. As a result, multiple versions of a class can safely coexist and interoperate during rolling upgrades, migrations, or client refreshes.

Reviewer: ```suggestion
Hazelcast achieves this through Compact serialization, which stores object schemas as independent metadata and identifies each version using a unique 64-bit fingerprint. When an object changes, Hazelcast detects the new schema automatically, assigns it a distinct fingerprint, and propagates it across the cluster. As a result, multiple versions of a class can safely coexist and interoperate during rolling upgrades, migrations, or client refreshes. The new and the old schema is associated with the same object type(e.g. class in Java) based on the `type name` defined in the serializer. They use the same type names. The type name can be any unique string, e.g. fully qualified class name in Jave.
```

Author: I have fixed it!

> While most schema changes can be made safely, **backward incompatibility** may occur when a change prevents old readers from interpreting previously written data correctly. This typically happens in cases such as field renames (changing a field’s name rather than adding/removing it), narrowing a type (for example, `long` to `int`), altering a field’s semantic meaning, or modifying the partitioning key used for data routing. When backward incompatibility is introduced, migration strategies must be applied—such as deploying a temporary pipeline or migration job to transform existing data, introducing versioned maps (`Order_v1`, `Order_v2`), or scheduling controlled rollouts with data export and reimport. These approaches ensure data integrity and consistent behavior during cluster upgrades or schema transitions.

Reviewer: Can you explain why this would cause backward incompatiblity in terms of serialization? `or modifying the partitioning key used for data routing` did you mean incompatibility in the cluster where data is stored?

Author: Yes - I meant incompatibility in the cluster; the write up is about "schema evolution" so I felt that it's not just about map values that should be described.

| `Fingerprint`
| A 64-bit identifier derived from the schema definition using a link:https://en.wikipedia.org/wiki/Rabin_fingerprint[Rabin Fingerprint] algorithm. Any change to a schema produces a new fingerprint.

| `Backward Compatibility`
| The ability of newer code (reader) to correctly read data serialized by older code (writer).

| `Forward Compatibility`
| The ability of older code (reader) to correctly read data serialized by newer code (writer), ignoring unknown fields.

| `FieldKind`
| The internal representation of a field’s data type in Compact serialization (e.g., `INT32`, `STRING`, `DECIMAL`, `ARRAY_OF_INT64`).

|===

[NOTE]
====
The following information applies to cases where Compact serialization is used. There's no guarantee that it works when Java Serialization or other supported serialization mechanisms are used
====

== Core principles and strategies

Schema evolution in Hazelcast is guided by a set of core principles designed to ensure that data changes remain predictable, compatible, and easy to manage across rolling upgrades, mixed client versions, and persistent clusters.

=== Stable `typeName`

> Each Compact type is identified by a `typeName`, which must remain stable throughout the lifetime of a class.

Reviewer: ```suggestion
Each Compact type is identified by a `typeName`, which must remain the same throughout the lifetime of a class.
```

Maintaining a stable `typeName` allows new and old members to exchange data transparently, because Hazelcast treats each version of the same `typeName` as an evolution rather than a different entity.

=== Deterministic writes

Hazelcast derives a schema fingerprint from the fields written in the serializer’s `write` method.
The structure and order of these writes must be **deterministic** — meaning they do not depend on runtime conditions or object state.

Conditional writes, skipped fields, or dynamic field names can lead to multiple fingerprints being generated for logically identical objects, causing unexpected deserialization errors or redundant schema registrations.

To ensure deterministic writes:

* Always write all defined fields, even if their values are `null` or default.
* Avoid conditional or state-dependent logic inside the `write` method.
* Keep field names and their corresponding `FieldKind` stable between releases.

=== Schema evolution approaches

Hazelcast supports two primary approaches to managing schema evolution, depending on operational needs and compatibility requirements.

[cols="1,3", options="header"]
|===
| Approach | Description

> | **Single Map Evolution**

Reviewer: ```suggestion
| **Single IMap Evolution**
```

> | All schema changes occur within the same map and `typeName`. This approach allows members to coexist using different schema versions during rolling upgrades. It is suitable for additive or compatible changes (for example, adding new fields, widening types).

Reviewer: ```suggestion
| All schema changes occur within the same IMap and `typeName`. This approach allows members to coexist using different schema versions during rolling upgrades. It is suitable for additive or compatible changes (for example, adding new fields).
```
Widening types or removing a field is forward incompatible if the old reader does not do field existence check (which is usually not the case).

| **Versioned Maps**

> | Each schema version is stored in a separate map (for example, `Order_v1`, `Order_v2`). This pattern isolates incompatible changes such as field renames, type narrowing, or semantic changes. Data can then be migrated using a pipeline or map-to-map transformation before decommissioning the older map.

Reviewer: We should say that this would require a certain design for the user application. Hence, different from first option which does not require user app design change.

Author: fixed

This section explains when you can keep existing indexes, when you must recreate them, and how to handle indexes during migrations.

=== Compatible schema changes

Additive or type-widening changes (for example, adding `currency` or widening `int` → `bigint`) do not invalidate existing indexes on unchanged fields.

* Indexes on old fields continue to work.
* You may add new indexes for newly introduced fields at any time.

=== Incompatible schema changes

Backward-incompatible changes (field rename, removal, type narrowing, or new key structure) require new index definitions aligned with the new schema.

* **Field rename/removal:** drop the old index and create a new one on the new field.
* **Type narrowing:** avoid in place; migrate to a new map and define indexes there.
* **Partitioning/key change (e.g. composite key):** create indexes on the **new map** only.

[NOTE]
====
Indexes are scoped per map. Versioned maps (for example, `orders` and `orders_v3`) hold independent index metadata; define indexes separately for each.
Create these indexes on the **target map first** (so queries are efficient as data arrives).
====

== Caveats and warnings

> * Compact fingerprint is not guaranteed to be stable across different versions of the Compact protocol or across language versions. The schema compatibility is indeed tied to the Hazelcast cluster version.

Author: discussed and fixed

> * When persistence is enabled, Hot Restart stores schema definitions in the same directory, but version mismatches can also occur if the cluster version changed and a Compact protocol version has been upgraded.

Reviewer: ???? `Compact protocol version has been upgraded`

* Certain field kinds (e.g. ARRAY_OF_COMPACT, COMPACT) have restricted evolution semantics (cannot change element kind or switch between nullable/non-nullable).

> * With Compact serialization, when a field is missing, Hazelcast provides type-specific defaults (0, false, null). This is relevant in cases when fields are removed.

Reviewer: ```suggestion
* With Compact serialization, when a field is missing, Hazelcast provides type-specific defaults (0, false, null). This is relevant in cases when fields are added.
```
This default is coded by user in their serializer in the new version, right? hazelcast does automatically do it, it gives you a chance to check for existence of the field in the serializer only, rest is up to the user what to assign for that field value, right?
